addons/package_hierarchy/models/package_links.py

```diff
@@ -0,0 +1,218 @@
+    an unlink for the purposes of this model).
+
+    Links may be created by and associated with movelines.
+    This allows the resulting package hierachy to be checked before
+    the moves have been completed, allowing early warning of
+    potential constraint violation.
+    """
+
+    _name = "package.hierarchy.link"
+    _description = "Links for constructing package hierarchies"
+    _order = "id"
+
+    _parent_name = "parent_id"
+
+    name = fields.Char(index=True, store=True, compute="_compute_name")
+
+    parent_id = fields.Many2one(
+        "stock.quant.package", string="Parent package", ondelete="cascade", check_company=True
+    )
+    child_id = fields.Many2one(
+        "stock.quant.package",
+        string="Child package",
+        ondelete="cascade",
+        check_company=True,
+        required=True
+    )
+    move_line_ids = fields.Many2many(
+        "stock.move.line",
+        column1="link_id",
+        column2="move_line_id",
+        string="Stock Move Line",
+        ondelete="cascade",
+        check_company=True,
+    )
+    company_id = fields.Many2one("res.company", default=False)
+
+    def construct(self):
+        for parent_package, links in self.groupby("parent_id"):
+            links.child_id.write(
+                {"parent_id": parent_package.id if parent_package else False}
+            )
+
+    @api.model
+    def create_unlinks(self, packages, move_lines=False):
+        """Create links that remove packages from hierarchies."""
+        # Warm the postgres cache
+        packages.get_move_lines_of_children(aux_domain=[("id", "in", move_lines.ids)])
+        link_vals = []
+        # Create unlinks from higher level packages
+        for package in packages:
+            if package.parent_id:
+                mls = package.get_move_lines_of_children(aux_domain=[("id", "in", move_lines.ids)])
+                link_vals.append(
+                    {
+                        "parent_id": False,
+                        "child_id": package.id,
+                        "move_line_ids": [(6, 0, mls.ids if mls else False)],
+                    }
+                )
+
+        if link_vals:
+            return self.create(link_vals)
+        return self.browse()
+
+    @api.depends("parent_id", "child_id")
+    def _compute_name(self):
+        for link in self.filtered(lambda p: not isinstance(p.id, models.NewId)):
+            link.name = (
+                "Link %s and %s" % (link.parent_id.name, link.child_id.name)
+                if link.parent_id and link.child_id
+                else "Unlink parent of %s" % link.child_id.name
+            )
+
+    @api.constrains("parent_id", "child_id")
+    def constrain_links(self):
+        """
+        Find all the links related to the links in self through their moves/moves lines,
+        and call validate links to check that they don't violate any of the constraints.
+
+        NOTE: To be 'more correct' we should check against all links from all moves
```

Comment:
Not all of them, only move lines related to the packages in the moves lines and links in self.